File: framework/languages/dotnet/src/Zlink.Framework/Runtime/Locations/ZLinkAutoConnectPlanner.cs
namespace Zlink.Framework.Runtime.Locations;

/// <summary>
/// A local capability that participates in auto connect: what this node
/// advertises for one mesh and what it dials from. The auto-connect type
/// and role classify the local side only; remote MeshNode descriptors carry
/// no role — publishing a descriptor in a mesh namespace is what makes a
/// node a dial target of that mesh (40-location-runtime §5).
/// </summary>
internal sealed record ZLinkAutoConnectLocal(
    ZLinkLocationAutoConnectType AutoConnectType,
    string MeshName,
    ZLinkLocationRole Role,
    RoutingId? NodeRid,
    string Endpoint);

/// <summary>
/// One connect target chosen by the planner, keyed by the descriptor RID so
/// an endpoint change under the same key is treated as a handover.
/// </summary>
internal sealed record ZLinkAutoConnectTarget(
    string TargetKey,
    RoutingId NodeRid,
    string Endpoint,
    bool Draining = false,
    string? OwnerId = null,
    bool InitiatesSpotRouterLink = true);

/// <summary>
/// Pure desired-target-set computation over a mesh descriptor snapshot,
/// with the pairwise initiator order for symmetric meshes. No I/O, so every
/// rule is unit-testable in isolation.
/// </summary>
internal static class ZLinkAutoConnectPlanner
{
    internal static IReadOnlyDictionary<string, ZLinkAutoConnectTarget> ComputeDesired(
        ZLinkAutoConnectLocal local,
        IReadOnlyList<ZLinkMeshNodeDescriptor> descriptors)
    {
        var desired = new Dictionary<string, ZLinkAutoConnectTarget>(StringComparer.Ordinal);
        foreach (var descriptor in descriptors)
        {
            if (!string.Equals(descriptor.MeshName, local.MeshName, StringComparison.Ordinal)
                || string.IsNullOrEmpty(descriptor.Endpoint)
                || IsSelf(local, descriptor)
                || !ShouldDial(local, descriptor))
            {
                continue;
            }

            var target = new ZLinkAutoConnectTarget(
                descriptor.Rid.ToHex(),
                descriptor.Rid,
                descriptor.Endpoint,
                descriptor.Draining,
                descriptor.OwnerId,
                local.AutoConnectType != ZLinkLocationAutoConnectType.SpotMesh
                || LocalIsInitiator(local, descriptor));
            desired[target.TargetKey] = target;
        }

        return desired;
    }

    internal static int CountDiscoveredPeers(
        ZLinkAutoConnectLocal local,
        IReadOnlyList<ZLinkMeshNodeDescriptor> descriptors)
    {
        return descriptors.Count(descriptor =>
            string.Equals(descriptor.MeshName, local.MeshName, StringComparison.Ordinal)
            && !string.IsNullOrEmpty(descriptor.Endpoint)
            && !IsSelf(local, descriptor));
    }

    private static bool IsSelf(ZLinkAutoConnectLocal local, ZLinkMeshNodeDescriptor descriptor)
    {
        if (local.NodeRid is { Size: > 0 } localRid && localRid.Equals(descriptor.Rid))
        {
            return true;
        }

        return string.Equals(descriptor.Endpoint, local.Endpoint, StringComparison.Ordinal);
    }

    /// <summary>
    /// Dial decisions come from the local role alone. Advertise-only roles
    /// (server routers, publishers) never dial; dial-only roles (dealers,
    /// subscribers) dial every descriptor of the mesh; symmetric meshes use
    /// the pairwise initiator so a pair never double-connects — both sides
    /// dialing creates two links for one routing id and breaks
    /// rid-addressed requests.
    /// </summary>
    private static bool ShouldDial(ZLinkAutoConnectLocal local, ZLinkMeshNodeDescriptor descriptor) =>
        local.AutoConnectType switch
        {
            ZLinkLocationAutoConnectType.RouteMesh =>
                local.Role == ZLinkLocationRole.Router && LocalIsInitiator(local, descriptor),
            ZLinkLocationAutoConnectType.ClientServer => local.Role == ZLinkLocationRole.Dealer,
            ZLinkLocationAutoConnectType.DealerMesh =>
                local.Role == ZLinkLocationRole.Dealer && LocalIsInitiator(local, descriptor),
            ZLinkLocationAutoConnectType.Fanout => local.Role == ZLinkLocationRole.Sub,
            ZLinkLocationAutoConnectType.SpotMesh => local.Role == ZLinkLocationRole.Spot,
            _ => false
        };

    /// <summary>
    /// Pairwise initiator for symmetric meshes: compare routing ids by byte
    /// order when both exist, otherwise compare endpoints ordinally; only
    /// the smaller side dials so the pair never double-connects. A member
    /// with no endpoint cannot be dialed, so it always initiates toward
    /// dialable peers regardless of the order.
    /// </summary>
    private static bool LocalIsInitiator(ZLinkAutoConnectLocal local, ZLinkMeshNodeDescriptor descriptor)
    {
        if (string.IsNullOrEmpty(local.Endpoint))
        {
            return true;
        }

        if (local.NodeRid is { Size: > 0 } localRid)
        {
            var byRid = string.CompareOrdinal(localRid.ToHex(), descriptor.Rid.ToHex());
            if (byRid != 0)
            {
                return byRid < 0;
            }
        }

        return string.CompareOrdinal(local.Endpoint, descriptor.Endpoint) < 0;
    }
}

File: framework/languages/java/zlink-framework-core/src/main/java/systems/zlink/framework/runtime/locations/ZLinkAutoConnectPlanner.java
package systems.zlink.framework.runtime.locations;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import systems.zlink.contracts.core.RoutingId;
import systems.zlink.framework.locations.ZLinkLocationAutoConnectType;
import systems.zlink.framework.locations.ZLinkLocationRole;
import systems.zlink.framework.locations.ZLinkPeerLocation;

final class ZLinkAutoConnectPlanner {
    private ZLinkAutoConnectPlanner() {
    }

    record Local(
        ZLinkLocationAutoConnectType type,
        String meshName,
        ZLinkLocationRole role,
        RoutingId nodeRid,
        String endpoint) {
    }

    record Target(
        String key,
        RoutingId nodeRid,
        ZLinkLocationRole role,
        String endpoint,
        Map<String, String> metadata,
        String ownerId) {
    }

    static boolean isRoleAllowed(
        ZLinkLocationAutoConnectType type,
        ZLinkLocationRole role) {
        return switch (type) {
            case ROUTE_MESH -> role == ZLinkLocationRole.ROUTER;
            case CLIENT_SERVER -> role == ZLinkLocationRole.ROUTER
                || role == ZLinkLocationRole.DEALER;
            case DEALER_MESH -> role == ZLinkLocationRole.DEALER;
            case FANOUT -> role == ZLinkLocationRole.PUB || role == ZLinkLocationRole.SUB;
            case SPOT_MESH -> role == ZLinkLocationRole.SPOT
                || role == ZLinkLocationRole.ROUTER;
            default -> false;
        };
    }

    static Map<String, Target> computeDesired(Local local, List<ZLinkPeerLocation> peers) {
        Map<String, Target> desired = new HashMap<>();
        for (ZLinkPeerLocation peer : peers) {
            if (peer.autoConnectType() != local.type()
                || !peer.meshName().equals(local.meshName())
                || !isRoleAllowed(local.type(), peer.role())
                || peer.endpoint() == null
                || peer.endpoint().isBlank()) {
                continue;
            }
            if (isSelf(local, peer) || !shouldDial(local, peer)) {
                continue;
            }
            Target target = new Target(
                targetKeyOf(peer),
                peer.nodeRid(),
                peer.role(),
                peer.endpoint(),
                peer.metadata(),
                peer.ownerId());
            desired.put(target.key(), target);
        }
        return desired;
    }

    private static String targetKeyOf(ZLinkPeerLocation peer) {
        String identity = hasRid(peer.nodeRid())
            ? peer.nodeRid().toHex()
            : peer.endpoint();
        return peer.role().name().toLowerCase(java.util.Locale.ROOT) + "|" + identity;
    }

    private static boolean isSelf(Local local, ZLinkPeerLocation peer) {
        if (hasRid(local.nodeRid()) && hasRid(peer.nodeRid())
            && local.nodeRid().equals(peer.nodeRid())) {
            return true;
        }
        return peer.endpoint().equals(local.endpoint());
    }

    private static boolean shouldDial(Local local, ZLinkPeerLocation peer) {
        return switch (local.type()) {
            case ROUTE_MESH -> local.role() == ZLinkLocationRole.ROUTER
                && peer.role() == ZLinkLocationRole.ROUTER
                && localIsInitiator(local, peer);
            case CLIENT_SERVER -> local.role() == ZLinkLocationRole.DEALER
                && peer.role() == ZLinkLocationRole.ROUTER;
            case DEALER_MESH -> local.role() == ZLinkLocationRole.DEALER
                && peer.role() == ZLinkLocationRole.DEALER
                && localIsInitiator(local, peer);
            case FANOUT -> local.role() == ZLinkLocationRole.SUB
                && peer.role() == ZLinkLocationRole.PUB;
            case SPOT_MESH -> local.role() == ZLinkLocationRole.SPOT
                && peer.role() == ZLinkLocationRole.SPOT;
            default -> false;
        };
    }

    private static boolean localIsInitiator(Local local, ZLinkPeerLocation peer) {
        if (local.endpoint() == null || local.endpoint().isBlank()) {
            return true;
        }
        if (hasRid(local.nodeRid()) && hasRid(peer.nodeRid())) {
            int byRid = local.nodeRid().toHex().compareTo(peer.nodeRid().toHex());
            if (byRid != 0) {
                return byRid < 0;
            }
        }
        return local.endpoint().compareTo(peer.endpoint()) < 0;
    }

    static boolean hasRid(RoutingId rid) {
        return rid != null && rid.size() > 0;
    }
}
